Generate Locations OCI Image

joaopapereira opened the pull request:
A couple of things had to change in order for this change to be done:
- Change the way the progress bar is being used, please refer to the first commit message for more information
- Add Logger with the level set, this will allow us to display messages with different level, not essential for this work, but helpful for the future
- Work to create the Locations OCI Image

What is missing:
- [x] Add more test coverage
- [x] Check if we can move the locations code into the Bundle struct
- [x] Check if it makes sense to refactor the singleLayerReader, we duplicated it to start
- [ ] General refactoring
- [x] The UI is being passed around, not sure if we want that...

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,65 @@ import (
-	o.imagesLock.AddImageRef(ref)
+func (o *ImagesLock) AddImageRef(ref lockconfig.ImageRef, bundle bool) {
+	for _, image := range o.imagesRef {
+		if image.Image == ref.Image {
```
may have to override in case when isbundle==false before, and new bundle is true.

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/registry/with_progress.go:
```diff
@@ -0,0 +1,63 @@
+	return w.delegate.FirstImageExists(digests)
+}
+
+func (w *WithProgress) MultiWrite(imageOrIndexesToUpload map[regname.Reference]remote.Taggable, concurrency int) error {
```
instead of accessing private opts, lets change sig of MultiWrite to include optional remote options.

joaopapereira replied:
fixed

joaopapereira committed:
Prefactor to include ImagesRef on Bundle struct

This prefactor will add all the Images that are associated to the
current bundle to a map. This will include all the images from the
current bundle plus all the images from the nested bundles.

This information will be used in next commits to distinguish between
what which images are bundles and which ones are not

joaopapereira committed:
Remove println from test

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -18,6 +18,7 @@ import (
 )
 
 func (o *Bundle) AllImagesLock(concurrency int) (*ImagesLock, error) {
+	// TODO: See if we can add the locations here
```
per our discussion, this method should probably start returning Bundles that were found. we can still keep returning ImagesLock but we should probably rename it to ImageRefs since that's what it is at this point.

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -77,17 +84,63 @@ func (c CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error)
 		return nil, err
 	}
 
+	if c.originBundle != nil {
```
per our discussion, consolidating this functionality on the Bundle would nicely keep consumers of bundles oblivious to image locations.

joaopapereira replied:
fixed

joaopapereira committed:
Add the ability to fetch and save locations files

- Fetch the locations images if it can find them and uses the values,
  when not present it keeps old behavior

cppforlife commented on pkg/imgpkg/registry/with_progress.go:
```diff
@@ -0,0 +1,60 @@
+	w.logger.Start(uploadProgress)
+	defer w.logger.End()
+
+	return w.delegate.MultiWrite(imageOrIndexesToUpload, concurrency, append(opts, remote.WithProgress(uploadProgress))...)
```
i dunno if opts is being used at the call site, but since append is possibly mutating i would recommend `append(append([]remote.Option{}, opts...),  remote.WithProgress(uploadProgress))` to make sure we are copying it.

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/registry/registry.go:
```diff
@@ -107,9 +107,10 @@ func (r Registry) Image(ref regname.Reference) (regv1.Image, error) {
+func (r Registry) MultiWrite(imageOrIndexesToUpload map[regname.Reference]regremote.Taggable, concurrency int, opts ...regremote.Option) error {
 	return util.Retry(func() error {
-		return regremote.MultiWrite(imageOrIndexesToUpload, append(r.opts, regremote.WithJobs(concurrency))...)
+		lOpts := append(r.opts, opts...)
```
same as comment below about mutating opts

joaopapereira replied:
fixed

cppforlife commented on test/helpers/logger.go:
```diff
@@ -25,6 +25,10 @@ func (l Logger) Section(msg string, f func()) {
 	f()
 }
 
+func (l Logger) Warnf(msg string, args ...interface{}) {
```
do we need warnings in tests?

joaopapereira replied:
no, but we need to comply with `util.LoggerWithLevels` interface

cppforlife commented on pkg/imgpkg/cmd/copy.go:
```diff
@@ -124,14 +123,15 @@ func (c *CopyOptions) Run() error {
 			signatureRetriever = signature.NewNoop()
 		}
 
+		levelLogger.LogLevel = util.LogTrace
```
no longer needed?

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/cmd/copy.go:
```diff
@@ -327,8 +325,8 @@ func informUserToUseTheNonDistributableFlagWithDescriptors(logger Logger, includ
 
 	if includeNonDistributableFlag && noNonDistributableLayers {
-		logger.WriteStr("Warning: '--include-non-distributable-layers' flag provided, but no images contained a non-distributable layer.")
+		logger.Warnf("Warning: '--include-non-distributable-layers' flag provided, but no images contained a non-distributable layer.")
```
should Warnf add `Warning: `?

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,72 @@
+type ImageLocation struct {
+	Image string `json:"image"` // This generated yaml, but due to lib we need to use `json`
+	// Repository will not be used right now because all images will be in the bundle repository
+	Repository string `json:"repository,omitempty"` // This generated yaml, but due to lib we need to use `json`
```
should we remove Repository for now?

joaopapereira replied:
Done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,72 @@
+	Image string `json:"image"` // This generated yaml, but due to lib we need to use `json`
+	// Repository will not be used right now because all images will be in the bundle repository
+	Repository string `json:"repository,omitempty"` // This generated yaml, but due to lib we need to use `json`
+	Bundle     bool   `json:"bundle"`               // This generated yaml, but due to lib we need to use `json`
```
wdyt about calling this `IsBundle  bool  json:"isBundle"`. imho is makes it more understandable that this is a t/f kinda a thing vs some bundle configuration.

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,72 @@
+	LocationAPIVersion = "imgpkg.carvel.dev/v1alpha1"
+)
+
+type LocationConfig struct {
```
LocationConfig -> ImageLocationsConfig

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,72 @@
+
+	err := yaml.UnmarshalStrict(data, &lock)
+	if err != nil {
+		return lock, fmt.Errorf("Unmarshaling bundle lock: %s", err)
```
not a bundle lock

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,72 @@
+func NewLocationConfigFromBytes(data []byte) (LocationConfig, error) {
+	var lock LocationConfig
+
+	err := yaml.UnmarshalStrict(data, &lock)
```
hmm, whenever we add new fields, is "strict" going to screw us up?

joaopapereira replied:
This function does not allow unknown fields. This will only cause problems if you are using an older version of `imgpkg` with a newer version of the Image Locations configuration. To be fair I think we want that because when we reintroduce the Repository field we need to respect it.

cppforlife replied:
> This function does not allow unknown fields. This will only cause problems if you are using an older version of imgpkg with a newer version of the Image Locations configuration.

this may happen for if kapp-controller's impgkg is on version A, and bundle was created with imgpkg version A+1. i think if meaning of ImageLocations changes significantly we should rev apiVersion. if meaning doesnt change significantly then we should be ok with new fields being optional. wdyt?

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+	return &Locations{reader: reader, logger: logger}
+}
+
+func (r Locations) Fetch(registry image.ImagesMetadata, bundleRef name.Digest) (LocationConfig, error) {
```
should we move registry into a constructor?

joaopapereira replied:
Unfortunately, Fetch and Save need 2 different types of registry struct. One needs the Reader the other needs the Writer. If we moved the registry to the constructor we would have to provide a reader/writer to accommodate both cases.
The problem with this is that the Bundle struct currently only has Read access to the registry when collecting the Image References aka `Bundle.AllImagesRefs`.

I decided to move the registry to the function itself because the only situation we need to write on this struct is in the NoteCopy function, so we would be giving write access to the majority of the cases where we do not need it.

cppforlife replied:
sounds good. makes sense.

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+	return nil
+}
+
+func (r Locations) tagFromBundleRef(bundleRef name.Digest) (name.Tag, error) {
```
tagFromBundleRef -> locationsRefFromBundleRef

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+
+func (r Locations) Fetch(registry image.ImagesMetadata, bundleRef name.Digest) (LocationConfig, error) {
+	r.logger.Tracef("fetching Locations OCI Images for bundle: %s\n", bundleRef)
+	locTag, err := r.tagFromBundleRef(bundleRef)
```
locTag -> locRef (tag is more of an impl detail, but it's definitely a ref)

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+		return LocationConfig{}, fmt.Errorf("fetching location image: %s", err)
+	}
+
+	r.logger.Tracef("reading the Locations configuration file\n")
```
should Tracef require to add newline everywhere?

joaopapereira replied:
Not a fan of it, nevertheless I was trying to keep some consistency with the `Printf` function. But I am open to change all the logger functions to provide the newline on them.

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpDir)
```
im pretty sure it's erroring because directory not empty. os.RemoveAll?

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+		return fmt.Errorf("calculating locations image tag: %s", err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "locations")
```
lets name it imgpkg-bundle-locations to make it easier for folks to know who creates temp dirs if some get left around.

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,175 @@
+func (r Locations) Save(reg ImagesMetadataWriter, bundleRef name.Digest, config LocationConfig, ui ui.UI) error {
+	r.logger.Tracef("saving Locations OCI Image for bundle: %s\n", bundleRef.Name())
+
+	tag, err := r.tagFromBundleRef(bundleRef)
```
tag -> locRef

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,94 @@ import (
+	lockconfig.ImageRef
+	IsBundle bool
+}
+type ImagesRef []ImageRef
```
ImagesRef -> ImageRefs (plural of ImageRef, a single item in a collection). in other words, it's not a single Ref to bunch of images, it's multiple ImageRefs.

joaopapereira replied:
Done

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,94 @@ import (
+		}
+
+		if !found {
+			result = append(result, ref)
```
this is getting a bit funky with reassigning i pointer. to make it a bit less fragile, lets switch to

```
type ImageRefs struct {
  refs []ImageRef
}
```

joaopapereira replied:
I did this, the only funky thing is that now you need to call `ImageRefs.ImageRefs` to get the refs instead of just iterating over the `ImagesRefs` Slice

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,94 @@ import (
+	}
+	*i = result
+}
+func (i *ImagesRef) ImageRef(ref string) (ImageRef, bool) {
```
ImageRef -> Find

joaopapereira replied:
Done

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,94 @@ import (
 
-func (o ImagesLock) ImageRefs() []lockconfig.ImageRef {
-	return o.imagesLock.Images
+func (o ImagesLock) ImagesRefs() ImagesRef {
```
ImagesRefs -> ImageRefs per pluralization above.

joaopapereira replied:
Done

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,94 @@ import (
+		for j, imageRef := range result {
+			if imageRef.Image == ref.Image {
+				found = true
+				result[j] = ref
```
should we do a deep copy?

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,94 @@ import (
 
 type ImagesLock struct {
-	imagesLock   lockconfig.ImagesLock
+	imagesRef    []ImageRef
```
should we use ImageRefs type here?

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -67,24 +97,78 @@ func (o *Bundle) buildAllImagesLock(throttleReq *util.Throttle, processedImgs *p
+		panic(fmt.Sprintf("Internal inconsistency: The Bundle Reference '%s' does not have a digest", o.DigestRef()))
+	}
+
+	var imageRefs ImagesRef
+	locationsConfig, err := NewLocations(logger).Fetch(o.imgRetriever, bundleDigestRef)
+	if err == nil {
+		imageRefs = o.processLocations(resultImagesLock, locationsConfig)
+	} else if _, ok := err.(*LocationsNotFound); !ok {
+		return nil, err
+	} else {
+		imagesLock, err := o.imagesLockReader.Read(img)
+		if err != nil {
 			return nil, err
 		}
+
+		currentImagesLock := NewImagesLock(imagesLock, o.imgRetriever, o.Repo())
+		imageRefs = currentImagesLock.ImagesRefs()
+	}
```
i think we should always read ImagesLock (it's still source of truth for referenced images by the bundle), but if we also find locations then we should AddLocation to the image ref.

joaopapereira replied:
Done

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -9,53 +9,83 @@ import (
 		image := image.DeepCopy()
 		go func() {
-			imgsLock, imgRef, err := o.imagesLockIfIsBundle(throttleReq, image, processedImgs)
+			lNestedBundles, imgsLock, imgRef, err := o.imagesLockIfIsBundle(throttleReq, image, processedImgs, logger)
```
what's `l` in `lNestedBundles`? should we just call it nestedBundles?

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/bundle.go:
```diff
@@ -29,24 +31,83 @@ type Bundle struct {
 func (o *Bundle) Repo() string      { return o.plainImg.Repo() }
 func (o *Bundle) Tag() string       { return o.plainImg.Tag() }
 
+func (o *Bundle) AddImagesRef(refs ...ImageRef) {
```
given all the work in this PR, im not sure i see the purpose of imageRefs on bundle being public. from what i see it's only really used by the NoteCopy.

joaopapereira replied:
We are using it for testing to ensure that the found references are the ones we are expecting

cppforlife commented on pkg/imgpkg/bundle/bundle.go:
```diff
@@ -29,24 +31,83 @@ type Bundle struct {
+		}
+	}
+
+	originBundleDigestRef, err := regname.NewDigest(o.DigestRef())
+	if err != nil {
+		return err
+	}
+	destinationRef, err := regname.ParseReference(bundleProcessedImage.DigestRef)
+	if err != nil {
+		return err
+	}
+
+	destDigestRef := destinationRef.Context().Digest(originBundleDigestRef.DigestStr())
+
```
isnt ref that we care about here the bundleProcessedImage.DigestRef? why do we need to use digest from o.DigestRef? (two should be the same).

joaopapereira replied:
yes you are right

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -113,70 +125,72 @@ func (c CopyRepoSrc) getSourceImages() (*ctlimgset.UnprocessedImageRefs, error)
-
-	imgLock, err := bundle.AllImagesLock(c.Concurrency)
+func (c *CopyRepoSrc) getBundleImageRefs(bundleRef string) ([]*ctlbundle.Bundle, ctlbundle.ImagesRef, error) {
+	c.originBundle = ctlbundle.NewBundle(bundleRef, c.registry)
```
instead of saving this off, we should just return rootBundle. looks like `...) (*Bundle, []*ctlbundle.Bundle, ctlbundle.ImagesRef, error)`. otherwise it's just a hidden contract.

joaopapereira replied:
fixed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -77,34 +82,41 @@ func (c CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error)
 		return nil, err
 	}
 
+	for _, bundle := range bundles {
```
shouldnt bundles also contain `originBundle` so that it can get a NoteCopy call?

joaopapereira replied:
the originBundle is added as part of the call to `Bundle.AllImagesRefs`

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -9,53 +9,83 @@ import (
+		return nil, ImagesRef{}, err
+	}
+
+	imagesRef := imagesLock.ImagesRefs()
+	// Ensure that the correct IsBundle flag is provided.
+	// This loop needs to happen because we skipped some images for some bundle, and only at this point we have
+	// the full list of ImagesRef created and can fill the gaps inside each bundle
+	for _, bundle := range bundles {
+		for _, ref := range bundle.ImagesRef() {
+			imgRef, found := imagesRef.ImageRef(ref.Image)
+			if !found {
+				panic(fmt.Sprintf("Internal inconsistency: The Image '%s' cannot be found in the total list of images", ref.Image))
+			}
+
+			bundle.AddImagesRef(imgRef)
+		}
+	}
```
i think we need to figure out a better way to do this. trying to think what are we actually storing the state for...

joaopapereira replied:
We are storing the state for each bundle of the Images that are part of it. This later is used when we are trying to generate each individual Locations OCI Image to find from the list of the ProcessedImages the ones that each bundle needs.

joaopapereira committed:
Refactoring ImagesRef -> ImageRefs

- Change the struct ImagesRef to be a struct instead of a name type

joaopapereira committed:
Use ImagesLock for each bundle as the base for locations

Always read the ImagesLock and only add the new location and if the
image is a bundle or not from the Locations OCI Image

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -13,52 +13,103 @@ import (
+
+	return ImageRef{}, false
+}
+func (i ImageRefs) ImageRefs() []ImageRef {
```
if you do not like ImageRefs{}.ImageRefs(), it could be called ImageRefs{}.All()

joaopapereira committed:
Make imageRefs functions private

- correct some typos
- correct issues found in the processLocation function

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -9,82 +9,165 @@ import (
-					errChan <- fmt.Errorf("Merging images for bundle '%s': %s", image.Image, err)
-					return
-				}
+			if nestedBundles != nil {
```
dont think we need to check on nil here. it's typed so ... will take care of it.

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -9,82 +9,165 @@ import (
+			// Adds Image to the resulting ImagesLock
+			processedImageRefs.AddImagesRef(ImageRef{
+				ImageRef: imgRef,
+				IsBundle: nestedBundles != nil, // nestedBundles will be != nil when the image is a bundle
```
we should check `len(nestedBundles) > 0` instead since nil and empty are not same.

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -59,12 +59,14 @@ func (c CopyRepoSrc) CopyToTar(dstPath string) error {
 
-func (c CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error) {
-	unprocessedImageRefs, err := c.getSourceImages()
+func (c *CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error) {
```
do we need to make CopyRepoSrc a pointer here after removal of originBundle?

joaopapereira replied:
removed

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -59,12 +59,14 @@ func (c CopyRepoSrc) CopyToTar(dstPath string) error {
-func (c CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error) {
-	unprocessedImageRefs, err := c.getSourceImages()
+func (c *CopyRepoSrc) CopyToRepo(repo string) (*ctlimgset.ProcessedImages, error) {
+	c.logger.Tracef(fmt.Sprintf("CopyToRepo(%s)\n", repo))
```
fmt.Sprintf is not necessary given f in Tracef

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations.go:
```diff
@@ -0,0 +1,176 @@
+)
+
+const (
+	locationsTag string = "%s-%s.locations.imgpkg"
```
lets changing this to `.image-locations.imgpkg`

cppforlife replied:
locationsTag -> locationsTagFmt

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,70 @@
+
+const (
+	LocationFilepath   = "images-locations.yml"
+	LocationKind       = "Locations"
```
`LocationKind       = "Locations"`
to
`ImageLocationsKind       = "ImageLocations"`

joaopapereira replied:
Done

cppforlife commented on pkg/imgpkg/bundle/locations_config.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+const (
+	LocationFilepath   = "images-locations.yml"
```
"images-locations.yml" -> "image-locations.yml"

joaopapereira replied:
done

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -9,82 +9,165 @@ import (
+	resultImageRefs := imageRefs.DeepCopy()
+	processedImageRefs := ImageRefs{}
+	for _, imgRef := range imageRefs.ImageRefs() {
+		for _, image := range locationsConfig.Images {
```
image -> imgLoc?

joaopapereira replied:
done

joaopapereira committed:
Change ImageLocation.IsBundle to become a pointer

joaopapereira committed:
Multiple fixes from PR comment

joaopapereira commented:
> > This function does not allow unknown fields. This will only cause problems if you are using an older version of imgpkg with a newer version of the Image Locations configuration.
>
> this may happen for if kapp-controller's impgkg is on version A, and bundle was created with imgpkg version A+1. i think if meaning of ImageLocations changes significantly we should rev apiVersion. if meaning doesnt change significantly then we should be ok with new fields being optional. wdyt?

This is exactly the scenario that we want to guard against. In your example, `imgpkg` assumes that the images will be present in the bundle repository, which might not be true. This will cause an error because kapp-controller will not be able to get the Images that are part of the bundle
